### DISCUSSION TITLE
Pytest error: You tried to access the function scoped fixture "browser" with a session scoped request object

### DISCUSSION THREAD

**Discussion by amanueltrutag** [pytest-dev/pytest#10667] [Q&A]
Hello all,
I was trying to use a session based test fixture for selenium tests but I am getting the error message :

 "_You tried to access the function scoped fixture browser with a session scoped request object_".  

 It works fine if the fixture scope is defined as "function" but then it will keep initializing the webdriver for every test module. I wanted to have it initialized once, and pass it to all the test functions in the different test modules. Below is the code:

**Contents of my Conftest.py file**

```
@pytest.fixture(scope="session", autouse=True)
def test_obj(base_url, browser, browser_version, testname):
    "Return an instance of Base Page that knows about the third party integrations"
    try:

        test_obj = PageFactory.get_page_object("Home Page", base_url=base_url)
        test_obj.set_calling_module(testname)

        # Setup and register a driver
        test_obj.register_driver(browser, browser_version)
        yield test_obj
        # Teardown
        test_obj.wait(3)
        test_obj.teardown()

    except Exception as e:
        print(f"Exception when trying to run test: {__file__}")
        print(f"Python says:{str(e)}")


@pytest.fixture(scope="session", autouse=True)
def testname(request):
    "pytest fixture for testname"
    try:
        name_of_test = request.node.name
        name_of_test = name_of_test.split("[")[0]

        return name_of_test

    except Exception as e:
        print(f"Exception when trying to run test: {__file__}")
        print(f"Python says:{str(e)}")


@pytest.fixture(scope="session", autouse=True)
def browser(request):
    "pytest fixture for browser"
    try:
        return request.config.getoption("--browser")

    except Exception as e:
        print(f"Exception when trying to run test: {__file__}")
        print(f"Python says:{str(e)}")


@pytest.fixture(scope="session", autouse=True)
def base_url(request):
    "pytest fixture for base url"
    try:
        return request.config.getoption("--web_url")

    except Exception as e:
        print(f"Exception when trying to run test: {__file__}")
        print(f"Python says:{str(e)}")

@pytest.fixture(scope="session", autouse=True)
def browser_version(request):
    "pytest fixture for browser version"
    try:
        return request.config.getoption("--ver")

    except Exception as e:
        print(f"Exception when trying to run test: {__file__}")
        print(f"Python says:{str(e)}")

def pytest_addoption(parser):
    "Method to add the option to ini."
    try:

        parser.addoption(
            "--browser",
            dest="browser",
            action="append",
            help="Browser. Valid options are firefox, edge and chrome",
        )
        parser.addoption(
            "--web_url",
            dest="url",
            action="append",
            default=base_url_conf.base_url,
            help="The url of the application",
        )
        parser.addoption(
            "--ver",
            dest="browser_version",
            action="append",
            help="The version of the browser: a whole number",
        )

    except Exception as e:
        print(f"Exception when trying to run test: {__file__}")
        print(f"Python says:{str(e)}")
```
**Contents of my test_1_login_page.py**
```
def test_login_with_invalid_credentials(test_obj):

    base_url = conf.base_url
    invalid_login_text = test_data.invalid_login_text
    user_name = "unknownuser@website.com"
    password = "unknown"

    # Turn on the highlight feature
    test_obj.turn_on_highlight()

    # set the user name
    result_flag = test_obj.set_username(user_name)

    # set the password
    if result_flag:
        result_flag = test_obj.set_password(password)

    # Click the submit button
    if result_flag:
        result_flag = test_obj.click_submit_button()

    # Check if we successfully logged in

    login_text = test_obj.get_invalid_login_text()

    assert login_text == invalid_login_text
```

when running pytest, I call it from the command line and pass the parameters there as shown:
```
pytest tests\test_1_login_page.py --browser chrome --ver 108
```
but I get the error message:
```
________________________________ ERROR at setup of test_login_with_valid_credentials[chrome] _________________________________ 
ScopeMismatch: You tried to access the function scoped fixture browser with a session scoped request object, involved factories:
conftest.py:10:  def test_obj(base_url, browser, browser_version, testname)
===================================================== 2 errors in 0.04s ====================================================== 
```
 Any suggestions on what I can fix this?

**Comment by arthuRHD** [comment]
Hi @amanueltrutag 

It could be many things.

1. Are you sure there isn't another fixture name `browser` from a dependency ?
2. You're using `autouse=True` in every fixtures. `autouse` is for fixtures that doesn't need to be requested in order to be executed but you actually request `browser`, `base_url`, `testname` and `browser_version` inside your `test_obj`. In my opinion, only the `autouse` on `test_obj` is useful. [More on this](https://pytest.org/en/latest/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-requesthttps://pytest.org/en/latest/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-request). Maybe this is the root of your problem I don't know, keep me updated 😉 

### Extra tips
- Avoid naming fixtures with `test_*`, It can be misunderstood as a testcase.
- Don't reinvent the wheel, check out the [*Get Started*](https://pytest-selenium.readthedocs.io/en/latest/user_guide.html#quick-start) of the package.
- Check out pytest hooks just like you did with [`pytest_addoption`](https://docs.pytest.org/en/6.2.x/reference.html?highlight=hook#initialization-hooks). There is a [`pytest_configure`](https://docs.pytest.org/en/6.2.x/reference.html?highlight=hook#pytest.hookspec.pytest_configure) hook that you can override to directly retrieve the [`config`](https://docs.pytest.org/en/6.2.x/reference.html?highlight=hook#pytest.config.Config) argument instead of using the fixture [`request`](https://docs.pytest.org/en/6.2.x/reference.html?highlight=hook#request). This hook is executed before fixture detection and after `pytest_addoption`.

**Reply by amanueltrutag** [reply]
Hi @arthuRHD,

Thank you for your detailed reply. I really appreciate it. So after fighting with it for a while, I figured out it was something to do with my `def pytest_addoption(parser):` function. I don't know precisely why it's doing that, but after removing the `def pytest_addoption(parser):`  function and all the fixtures associated with it, i.e., `def browser(request):` ,`def base_url(request):`  and `def browser_version(request):`, it started working as I expected it. I was using the command line options to pass the specific browser I wanted to use for the test. But after the above battle, I decided to parametrize the list of browsers instead of selecting it at the command line. So my updated **conftest.py** file looks something like this:


```
@pytest.fixture(scope="session", params=['chrome','edge','firefox'])
def test_obj(request):
    "Return an instance of Base Page that knows about the third party integrations"
    try:

        test_obj = PageFactory.get_page_object("home page")

        # Setup and register a driver
        test_obj.register_driver(request.param)
        
        yield test_obj
        # Teardown
        test_obj.wait(3)
        test_obj.teardown()

    except Exception as e:
        print(f"Exception when trying to run test: {__file__}")
        print(f"Python says:{str(e)}")


@pytest.hookimpl(hookwrapper=True)
def pytest_runtest_makereport(item, call):
    pytest_html = item.config.pluginmanager.getplugin("html")
    outcome = yield
    report = outcome.get_result()

    extra = getattr(report, "extra", [])
    if report.when == "call":

        feature_request = item.funcargs["request"]
        testObj = feature_request.getfixturevalue("test_obj")
        testName = item.name.split("[")[0]
        screenshot_file = testObj.save_screenshot(testName)
        extra.append(pytest_html.extras.image(screenshot_file))
        xfail = hasattr(report, "wasxfail")

        if (report.skipped and xfail) or (report.failed and not xfail):
            # only add additional html on failure
            extra.append(pytest_html.extras.html("<div>Additional HTML</div>"))
        report.extra = extra
```

I did not make any changes to my **test_1_login_page.py**. Its still the same as the original post.  Again, thank you for your inputs.

**Reply by tacerus** [reply]
Thanks a lot! I had the same problem, and totally missed it was coming from `pytest_addoption()` (though after knowing it, the argument reference in the error message does make sense).
In my case, the `scope` needed to be raised for the command line parameter as well:

```
def pytest_addoption(parser):
    parser.addoption('--target', action='store')

def pytest_generate_tests(metafunc):
    value = metafunc.config.option.target
    if 'target' in metafunc.fixturenames and value is not None:
        metafunc.parametrize('target', [value], scope='session')
```

`scope='session'` was missing here.
